### PR TITLE
Can't drag the cursor with the mouse

### LINKED ISSUE #102928: [Desktop]TextField cursor can be dragged with a mouse
On mobile, you can tap and drag the cursor, but it seems like that behavior hasn't been disabled when using a mouse:

out

## Steps to reproduce

  1. Run any app with a TextField, such as the one given below, on a device with a mouse.
  2. Paste or type some text into the field.
  3. Click and drag directly on top of the cursor.

Expected: A selection begins.
Actual: The cursor moves along with the mouse.

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Flutter Demo'),
        ),
        body: const Center(
          child: TextField(),
        ),
      ),
    );
  }
}

```

</details>

### PR DESCRIPTION
**Pull request by justinmc**
Previously, it was possible to click and drag the text selection cursor with the mouse.  This should only be possible on mobile with touch gestures.

| Old | New |
| --- | --- |
| out | out |

Fixes https://github.com/flutter/flutter/issues/102928

### CODE REVIEW

File: packages/flutter/test/widgets/text_selection_test.dart [hunk 1]
```diff
@@ -652,6 +653,54 @@ void main() {
...
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: EditableText(
+            controller: controller,
+            focusNode: FocusNode(),
+            backgroundCursorColor: Colors.white,
+            cursorColor: Colors.white,
+            style: const TextStyle(),
+            //selectionControls: CustomTextSelectionControls(),
```
Comment (nilsreichardt) [inline packages/flutter/test/widgets/text_selection_test.dart:671]: Is the command there for a purpose?

Comment (justinmc) [reply packages/flutter/test/widgets/text_selection_test.dart:671]: Nope I forgot it there, thanks for catching that.

**Comment by justinmc** [comment]
@chunhtai I bisected this bug to your PR https://github.com/flutter/flutter/pull/98073.  What do you think of this solution?

The analyzer failure seems messed up, I'll push a merge commit when master gets updated and see if that fixes it.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 2]
```diff
@@ -359,7 +359,18 @@ class TextSelectionOverlay {
...
-    _selectionOverlay.showHandles();
+    // Desktop platforms don't have selection handles.
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.macOS:
+      case TargetPlatform.fuchsia:
+      case TargetPlatform.linux:
+      case TargetPlatform.windows:
+        break;
+      case TargetPlatform.android:
+      case TargetPlatform.iOS:
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/text_selection.dart:370]: what about iOS with blue tooth mouse?

**Review by chunhtai — COMMENTED** [review]
This seems to be a weird approach to disable mouse drag, I was wondering whether you can fix this by making `_SelectionHandleOverlay` to only accept touch gesture?

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/text_selection.dart:370]: Good point, I should actually try that and see what happens.  I'm not sure how it handles showing the handles or not.

**Comment by justinmc** [comment]
The problem is that the underlying field needs to receive that drag gesture when using a mouse, so I would need to write a custom GestureDetector that receives touch drags but ignores mouse drags.

My thought was that we shouldn't be building this invisible handle there anyway, but after looking into it my current way isn't the right way to do that.  The place to decide if a handle should be built or not is TextSelectionControls, but I don't think I can fix it there without a breaking change.

I'll investigate the touch-only GestureDetector and try to do that.  Maybe I can look into a deeper refactor and not drawing the handle at all after my ContextMenu changes.

**Comment by justinmc** [comment]
@chunhtai I've implemented the touch-only gesture detector solution, ready for re-review.

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/text_selection.dart]: With the new solution, mobile platforms with a mouse will behave the same as desktop platforms with a mouse (clicking and dragging with a mouse will select, not drag the cursor).

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 3]
```diff
@@ -1208,12 +1208,25 @@ class _SelectionHandleOverlayState extends State<_SelectionHandleOverlay> with S
...
-            onPanStart: widget.onSelectionHandleDragStart,
-            onPanUpdate: widget.onSelectionHandleDragUpdate,
-            onPanEnd: widget.onSelectionHandleDragEnd,
+            gestures: <Type, GestureRecognizerFactory>{
+              PanGestureRecognizer: GestureRecognizerFactoryWithHandlers<PanGestureRecognizer>(
+                () => PanGestureRecognizer(
+                  debugOwner: this,
+                  // Selection handles can only be dragged by touch. Mouse
+                  // events on the handle will select the text behind.
+                  supportedDevices: <PointerDeviceKind>{ PointerDeviceKind.touch },
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/text_selection.dart:1219]: what about stylus?

and there is the wild card, PointerDeviceKind.unknown. as far as I know, adb comand and a11y command may use  `unknown`. I think a11y is probably fine to be excluded. adb commands are usually used during integration test, which may expect to drag the selection handle. I am not sure whether that would be a valid use case though

**Review by chunhtai — COMMENTED** [review]
Code looks good, just concern about the supported devices.

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 3 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/text_selection.dart:1219]: That makes sense, let me try adding `stylus` and `unknown` and make sure the tests still pass here.

Comment (justinmc) [reply packages/flutter/lib/src/widgets/text_selection.dart]: I added them plus separate tests for touch, stylus, and unknown (and the existing mouse test).  Ready for re-review.

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**